Add CI node checks and retries. Refactored the pipeline to resolve compilation errors and address incorrect syntax

leosomebird opened the pull request:
Refactored the pipeline to resolve compilation errors and address incorrect syntax

NISHIY-EKSDEE committed:
Add CI node checks and retries

NISHIY-EKSDEE committed:
Fix missing bracket

Commit:
Update Jenkinsfile

Commit:
Update node_checker.groovy

causten commented:
please update the title of the PR to indicate what is going on.

Commit:
Trigger CI rebuild

Commit:
Trigger CI rebuild

Commit:
Added LD_LIBRARY_PATH Env variables

Explicitly set LD_LIBRARY_PATH for docker.inside stages
to resolve "cannot open shared object file: libamdhip64.so"
errors during tests.

Commit:
Added debug code

Commit:
Layering the runtime on top of the existing image before launch cmakeBuild

The pipeline itself is healthy. The container it relies on is no longer shipping the HIP runtime, so anything that touches HIP at build‑time explodes.

Commit:
Install the package that actually ships the symlink

Commit:
Increased heartbeats

ghost commented:
Hi @causten @umangyadav. Is it possible to slightly increase the % coverage to allow the check to pass? Omkar suggested reaching out to Umang for that

umangyadav commented:
> Hi @causten @umangyadav. Is it possible to slightly increase the % coverage to allow the check to pass? Omkar suggested reaching out to Umang for that

Code coverage is not a required check.  we don't need to look code coverage for this PR.

Commit:
ci: trigger rebuild

Commit:
Merge branch 'patch-1' of https://github.com/leosomebird/rocMLIR into patch-1

Commit:
ci: trigger rebuild

Commit:
ci: trigger rebuild

Commit:
Update String dockerArgs()

the GPUs are never being passed through to the container that runs the navi3x leg.

umangyadav commented on mlir/utils/jenkins/Jenkinsfile:
```diff
@@ -1,1156 +1,1328 @@
+    return """
+      --privileged --ipc=host \
+      --device=/dev/kfd \
+      --device=/dev/dri/renderD128 --device=/dev/dri/renderD129 \
```
Why do you need to mount `renderD128/128` and `card0/1` ?

They may not be available on all the machines.

umangyadav replied:
> https://github.com/ROCm/rocMLIR/runs/41869741938T

This run failed with memory access fault. Likely due to some other reason but not because it can't find GPUs. 

Remove `/dev/dri/renderD128` etc. I don't think they would be necessary.

leosomebird replied:
Done, the build started.

umangyadav commented on mlir/utils/jenkins/Jenkinsfile:
```diff
@@ -1,1156 +1,1328 @@
+// helper
+def ensureHip() {
+    sh '''
+        if ! ldconfig -p | grep -q "libamdhip64.so$"; then
```
MLIR dockers will have `libamdhip64.so` by default. I don't see need to do this

leosomebird replied:
@umangyadav That's what I was thinking too. Here the test failed with `RuntimeError: failed to dlopen 'b'libamdhip64.so'': b'libamdhip64.so: cannot open shared object file: No such file or directory'` however https://dev.azure.com/ROCm-CI/ROCm-CI/_build/results?buildId=29003&view=logs&jobId=0393e6a6-c903-51f1-fdec-f985e037780e&j=0393e6a6-c903-51f1-fdec-f985e037780e&t=7803d640-52e4-5661-de44-cffe8326c9dc
Addomg this fixed it

umangyadav replied:
That is Azure pipelines they are not controlled byt this Jenkinsfile. 

Please remove `ensureHip()` calls. It is unnecessary.

umangyadav replied:
Azure pipeline code is here : https://github.com/ROCm/ROCm/blob/develop/.azuredevops/components/rocMLIR.yml

leosomebird replied:
Done, the build started.

umangyadav commented on mlir/utils/jenkins/Jenkinsfile:
```diff
@@ -1,1156 +1,1328 @@
+                                            ensureHip()
+
+                                            sh '''
+                                                # 1. Install the package that actually ships the symlink
+                                                if ! dpkg -s hip-devel-extras &>/dev/null; then
+                                                    apt-get update -qq
+                                                    DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
+                                                        hip-devel-extras
+                                                fi
+
+                                                # 2.  Fallback: create the symlink ourselves if the package layout changes again
+                                                for d in /opt/rocm/lib /opt/rocm/hip/lib; do
+                                                    if [ -d "$d" ] && [ ! -e "$d/libamdhip64.so" ]; then
+                                                        tgt=$(ls "$d"/libamdhip64.so.* 2>/dev/null | head -n1)
+                                                        [ -n "$tgt" ] && ln -s "$tgt" "$d/libamdhip64.so"
+                                                    fi
+                                                done
```
I don't see need to do this.  Docker container will everything setup properly.

leosomebird replied:
Mostly as a chip insurance against image drifts. We can remove it

umangyadav replied:
Please remove that.

leosomebird replied:
Done, the build started.

umangyadav commented on mlir/utils/jenkins/Jenkinsfile:
```diff
@@ -1,1156 +1,1328 @@
+                                    String requiredLabel = getLabelFromCodepath("${CODEPATH}")
+                                    String dockerImageName = dockerImage()
+                                    def dockerEnvVars = [
+                                        'PATH': "/opt/rocm/llvm/bin:/opt/rocm/bin:${env.PATH}", // Added /opt/rocm/bin
+                                        'HOME': '/home/jenkins',
+                                        'LD_LIBRARY_PATH': '/opt/rocm/lib:/opt/rocm/hip/lib:${env.LD_LIBRARY_PATH ?: ""}'
+                                    ]
```
Why is this necessary now ?

leosomebird replied:
Also as a workaround for the `libamdhip64.so` issue. If I remember correctly, it didn't help though, so I left it just in case of future issues. Le me know if I need to remove this

umangyadav replied:
Please remove it

leosomebird replied:
Done, the build started.

umangyadav commented on mlir/utils/jenkins/Jenkinsfile:
```diff
@@ -1,1156 +1,1328 @@
+                                                --test_dir ../mlir/test/xmir/bert-torch-tosa-e2e/ -o tuning_fusion_${ARCH}.tsv || echo 'Tuning fusion bert failed' """
+                                        }
+                                        sh 'rm -f build/CMakeCache.txt'
+                                        echo "===== Finished: Tune Fusion for ${ARCH} ====="
```
Is there a good reason why you need to add `echo` before and after each stage ?

leosomebird replied:
I can remove them, but they were a great help when I was figuring out what the pipeline does. Might be a good idea to leave them for future newcomers. Let me know if I need to remove it.

umangyadav replied:
I suggest removing it. We can look at blue ocean pipelines which shows nicely what it's running etc.

leosomebird replied:
Done, the build started.

umangyadav commented on mlir/utils/jenkins/node_checker.groovy:
```diff
@@ -0,0 +1,141 @@
+                // Run the health check *on the allocated node*
+                checkNodeHealth() // Throws NodeCheckException on failure
+
+                // If check passes, store name and break inner try to execute code
```
Can you add another step in node acquisition about doing gpu reset using 
`rocm-smi --gpureset -d 0` ?

leosomebird replied:
@umangyadav yes, will do.

umangyadav replied:
I can add this later : 
I tried it here : https://github.com/ROCm/rocMLIR/pull/1828 

It is failing with permission issues. You can try if you wish, if you know how to get around sudo permission issue.

If not, Let's get this merged first and i can add `--gpureset` later.

Commit:
Resolving comments

https://github.com/ROCm/rocMLIR/pull/1822#discussion_r2080344168
https://github.com/ROCm/rocMLIR/pull/1822#discussion_r2080346297
https://github.com/ROCm/rocMLIR/pull/1822#discussion_r2080353688
https://github.com/ROCm/rocMLIR/pull/1822#discussion_r2080354972
https://github.com/ROCm/rocMLIR/pull/1822#discussion_r2080355921

umangyadav reviewed (approved):
As a follow up PR we would also like to updat
https://github.com/ROCm/rocMLIR/blob/develop/mlir/utils/jenkins/Jenkinsfile.downstream 
https://github.com/ROCm/rocMLIR/blob/develop/mlir/utils/jenkins/Jenkinsfile.release
But let's get this merged first.

Commit:
Trigger build

Commit:
Trigger build

Commit:
Merge branch 'patch-1' of https://github.com/leosomebird/rocMLIR into patch-1

Commit:
Trigger build

ghost commented:
@umangyadav The tests are failing again. I believe this is indeed because the runtime can't find any GPU devices, same error `hipErrorNoDevice`. Should I get the direct mounting back and try again?
https://github.com/ROCm/rocMLIR/pull/1822/checks?check_run_id=42053406954

umangyadav commented:
> @umangyadav The tests are failing again. I believe this is indeed because the runtime can't find any GPU devices, same error `hipErrorNoDevice`. Should I get the direct mounting back and try again? https://github.com/ROCm/rocMLIR/pull/1822/checks?check_run_id=42053406954

Hi @leosomebird  We can give it a try. 
But that said code shouldn't just mount RenderD128/D129. 
But it should detect all available GPUs on the machines and mount all of them. 
For example, if the server has 8 GPUs then, we want to mount all of them. e.g. RenderD128/129/130/131/132/133/134/135
You can probably write some script to do that.

ghost commented:
> > @umangyadav The tests are failing again. I believe this is indeed because the runtime can't find any GPU devices, same error `hipErrorNoDevice`. Should I get the direct mounting back and try again? https://github.com/ROCm/rocMLIR/pull/1822/checks?check_run_id=42053406954
> 
> Hi @leosomebird We can give it a try. But that said code shouldn't just mount RenderD128/D129. But it should detect all available GPUs on the machines and mount all of them. For example, if the server has 8 GPUs then, we want to mount all of them. e.g. RenderD128/129/130/131/132/133/134/135 You can probably write some script to do that.

Good idea, there should be a way.
I'll work on that

Commit:
Update Jenkinsfile with dynamic device discovery

umangyadav commented:
@leosomebird changes to docker args not being reflected in Jenkins run for some reason. When i look at Jenkins pipeline logs i do not see all the render devices being mounted. I am not sure why is that though

Commit:
Included the function inside the Jenkinsfile, fixed errors

umangyadav commented:
@leosomebird  can this one be closed now ? https://github.com/ROCm/rocMLIR/pull/1835 is merged now

umangyadav commented:
Closing this one since https://github.com/ROCm/rocMLIR/pull/1835  superseded it.
